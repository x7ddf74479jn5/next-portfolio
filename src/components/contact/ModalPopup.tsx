import React, { useRef } from "react";
import FormButton from "src/components/contact/FormButton";
import { GridRow } from "src/components/layouts/Grid";
import useClickAway from "src/hooks/useClickAway";
import { useModalDispatch } from "src/hooks/useModalDispatch";
import styles from "src/styles/components/modal/ConfirmModal.module.scss";

import type { FormData } from "../../types/api/index";

export type ModalPopup = FormData & ModalPopupAction;
type ModalPopupAction = {
  cancel: () => void;
  apply: (formData: FormData) => Promise<unknown>;
};
type Props = {
  formData: FormData;
  cancel: ModalPopupAction["cancel"];
  apply: ModalPopupAction["apply"];
  isApplyButtonDisabled: boolean;
};

const ModalPopup: React.VFC<Props> = ({ formData, cancel, apply, isApplyButtonDisabled }) => {
  const ref = useRef(null);
  const { closeModal } = useModalDispatch();
  useClickAway(ref, closeModal);
  return (
    <div className="c-modal-popup" id="popup" ref={ref}>
      <div className={styles.modalPopupBody}>
        <h2 className={styles.modalPopupTitle} id="modal-title">
          入力内容の確認
        </h2>
        <div className={styles.modalPopupMain} id="popup-main">
          <div id="mail-content">
            <p>お名前：</p>
            <p>{formData.name}</p>
            <p>メールアドレス：</p>
            <p>{formData.email}</p>
            <p>お問い合わせ種別：</p>
            <p>{formData.category}</p>
            <p>お問い合わせ内容：</p>
            <p>{formData.description}</p>
          </div>
        </div>
        <GridRow>
          <FormButton
            buttonType="cancel"
            id="cancel-btn"
            type="button"
            onClick={() => {
              return cancel();
            }}
          >
            修正する
          </FormButton>
          <FormButton
            buttonType="apply"
            id="apply-btn"
            type="button"
            disabled={isApplyButtonDisabled}
            onClick={() => {
              return apply(formData);
            }}
          >
            送信する
          </FormButton>
        </GridRow>
      </div>
      {/* <ModalCloseButton onClick={closeModal} /> */}
    </div>
  );
};

export default ModalPopup;
