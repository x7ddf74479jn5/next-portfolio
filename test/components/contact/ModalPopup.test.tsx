import type { RenderResult } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ModalPopup from "src/components/contact/ModalPopup";
import { render } from "test/test-utils";

describe("ModalPopup", () => {
  let renderResult: RenderResult;
  let mockHandleApply = jest.fn();
  let mockHandleCancel = jest.fn();

  const formData = {
    name: "name",
    email: "email@example.com",
    category: "category",
    description: "description",
  };

  it("matches snapshot", () => {
    renderResult = render(
      <ModalPopup formData={formData} cancel={mockHandleCancel} apply={mockHandleApply} isApplyButtonDisabled={false} />
    );
    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it("matches snapshot", () => {
    renderResult = render(
      <ModalPopup formData={formData} cancel={mockHandleCancel} apply={mockHandleApply} isApplyButtonDisabled={true} />
    );
    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it("rendered correctly", () => {
    renderResult = render(
      <ModalPopup formData={formData} cancel={mockHandleCancel} apply={mockHandleApply} isApplyButtonDisabled={false} />
    );
    expect(renderResult.getByRole("heading")).toHaveTextContent("入力内容の確認");

    const containerNode = renderResult.container.querySelector("#mail-content");

    expect(containerNode).toHaveTextContent("お名前");
    expect(containerNode).toHaveTextContent("name");
    expect(containerNode).toHaveTextContent("メールアドレス");
    expect(containerNode).toHaveTextContent("email");
    expect(containerNode).toHaveTextContent("お問い合わせ種別");
    expect(containerNode).toHaveTextContent("category");
    expect(containerNode).toHaveTextContent("お問い合わせ内容");
    expect(containerNode).toHaveTextContent("description");

    const buttonElements = renderResult.getAllByRole("button");
    const cancelButton = buttonElements[0];
    const applyButton = buttonElements[1];
    const modalCloseButton = buttonElements[2];

    expect(buttonElements).toHaveLength(3);

    expect(cancelButton).toHaveAttribute("id", "cancel-btn");
    expect(cancelButton).toBeEnabled();
    expect(cancelButton).toHaveTextContent("修正する");

    expect(applyButton).toHaveAttribute("id", "apply-btn");
    expect(applyButton).toBeEnabled();
    expect(applyButton).toHaveTextContent("送信する");

    expect(modalCloseButton).toBeEnabled();
  });

  describe("user interaction", () => {
    let buttonElements: HTMLElement[];
    let cancelButton: HTMLElement;
    let applyButton: HTMLElement;
    describe("if apply button is not disabled", () => {
      beforeEach(() => {
        mockHandleApply = jest.fn();
        mockHandleCancel = jest.fn();
        renderResult = render(
          <ModalPopup
            formData={formData}
            cancel={mockHandleCancel}
            apply={mockHandleApply}
            isApplyButtonDisabled={false}
          />
        );

        buttonElements = renderResult.getAllByRole("button");
        cancelButton = buttonElements[0];
        applyButton = buttonElements[1];
      });

      it("should  apply if not disabled", () => {
        userEvent.click(applyButton);
        expect(mockHandleApply).toBeCalledWith(formData);
      });

      it("closes modal via cancel button", () => {
        userEvent.click(cancelButton);
        expect(mockHandleCancel).toBeCalled();
      });
    });

    it("if apply button is disabled", () => {
      renderResult = render(
        <ModalPopup
          formData={formData}
          cancel={mockHandleCancel}
          apply={mockHandleApply}
          isApplyButtonDisabled={true}
        />
      );

      buttonElements = renderResult.getAllByRole("button");
      cancelButton = buttonElements[0];
      applyButton = buttonElements[1];

      expect(applyButton).toBeDisabled();

      userEvent.click(applyButton);

      expect(mockHandleApply).not.toHaveBeenCalled();
    });
  });
});
