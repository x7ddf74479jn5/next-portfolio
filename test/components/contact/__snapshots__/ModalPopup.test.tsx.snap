// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalPopup matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="c-modal-popup"
    id="popup"
  >
    <div
      class="modalPopupBody"
    >
      <h2
        class="modalPopupTitle"
        id="modal-title"
      >
        入力内容の確認
      </h2>
      <div
        class="modalPopupMain"
        id="popup-main"
      >
        <div
          id="mail-content"
        >
          <p>
            お名前：
          </p>
          <p>
            name
          </p>
          <p>
            メールアドレス：
          </p>
          <p>
            email@example.com
          </p>
          <p>
            お問い合わせ種別：
          </p>
          <p>
            category
          </p>
          <p>
            お問い合わせ内容：
          </p>
          <p>
            description
          </p>
        </div>
      </div>
      <div
        class="gridRow"
      >
        <button
          class="btnCancel"
          id="cancel-btn"
          type="button"
        >
          修正する
        </button>
        <button
          class="btnApply"
          id="apply-btn"
          type="button"
        >
          送信する
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ModalPopup matches snapshot 2`] = `
<DocumentFragment>
  <div
    class="c-modal-popup"
    id="popup"
  >
    <div
      class="modalPopupBody"
    >
      <h2
        class="modalPopupTitle"
        id="modal-title"
      >
        入力内容の確認
      </h2>
      <div
        class="modalPopupMain"
        id="popup-main"
      >
        <div
          id="mail-content"
        >
          <p>
            お名前：
          </p>
          <p>
            name
          </p>
          <p>
            メールアドレス：
          </p>
          <p>
            email@example.com
          </p>
          <p>
            お問い合わせ種別：
          </p>
          <p>
            category
          </p>
          <p>
            お問い合わせ内容：
          </p>
          <p>
            description
          </p>
        </div>
      </div>
      <div
        class="gridRow"
      >
        <button
          class="btnCancel"
          id="cancel-btn"
          type="button"
        >
          修正する
        </button>
        <button
          class="btnApply"
          disabled=""
          id="apply-btn"
          type="button"
        >
          送信する
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
